subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultProjectComponentSelector.java

```diff
@@ -105,8 +107,8 @@ public String getProjectName() {
     public boolean matchesStrictly(ComponentIdentifier identifier) {
         assert identifier != null : "identifier cannot be null";
 
-        if (identifier instanceof DefaultProjectComponentIdentifier) {
-            DefaultProjectComponentIdentifier projectComponentIdentifier = (DefaultProjectComponentIdentifier) identifier;
+        if (identifier instanceof ProjectComponentIdentifier) {
```

Comment:
This new code is definitely less "safe" here in terms of casting. I wouldn't be against changing this back, but why special case checking for this internal type, where for other types like `Configuration` or `Project` we don't test for internal before casting?

Reply:
Any ideas why we do that elsewhere instead of checking for the type we're going to cast to?  I can't think of any benefits.  It probably won't break, as we wouldn't ever have the internal type not extend the public type, but it's still a little weird looking.

Reply:
I think the primary benefit is verbosity. If we had to check for `instanceof ProjectInternal` every place we casted to `ProjectInternal`, we'd have a lot more `instanceof` checks throughout the codebase. In practice, nobody is writing their own `Project` implementation, so this is "safe" to do so.